types/urllib/index.d.ts

```diff
@@ -0,0 +1,253 @@
+   */
+  rejectUnauthorized?: boolean;
+  /** A string or Buffer containing the private key, certificate and CA certs of the server in PFX or PKCS12 format. */
+  pfx?: string | Buffer;
+  /**
+   * A string or Buffer containing the private key of the client in PEM format.
+   * Notes: This is necessary only if using the client certificate authentication
+   */
+  key?: string | Buffer;
+  /**
+   * A string or Buffer containing the certificate key of the client in PEM format.
+   * Notes: This is necessary only if using the client certificate authentication
+   */
+  cert?: string | Buffer;
+  /** A string of passphrase for the private key or pfx. */
+  passphrase?: string;
+  /** A string describing the ciphers to use or exclude. */
+  ciphers?: string;
+  /** The SSL method to use, e.g.SSLv3_method to force SSL version 3. */
+  secureProtocol?: string;
+  /** follow HTTP 3xx responses as redirects. defaults to false. */
+  followRedirect?: boolean;
+  /** The maximum number of redirects to follow, defaults to 10. */
+  maxRedirects?: number;
+  /** Format the redirect url by your self. Default is url.resolve(from, to). */
+  formatRedirectUrl?: (a: any, b: any) => void;
+  /** Before request hook, you can change every thing here. */
+  beforeRequest?: (...args: any[]) => void;
+  /** let you get the res object when request connected, default false. alias customResponse */
+  streaming?: boolean;
+  /** Accept gzip response content and auto decode it, default is false. */
+  gzip?: boolean;
+  /** Enable timing or not, default is false. */
+  timing?: boolean;
+  /** Enable proxy request, default is false. */
+  enableProxy?: boolean;
+  /** proxy agent uri or options, default is null. */
+  proxy?: string | { [key: string]: any };
+}
+
+/**
+ * @param err Error
+ * @param data Outgoing message
+ * @param res http response
+ */
+export type Callback = (err: Error, data: any, res: http.IncomingMessage) => void;
+
+/**
+ * Handle all http request, both http and https support well.
+ *
+ * @example
+ * // GET http://httptest.cnodejs.net
+ * urllib.request('http://httptest.cnodejs.net/test/get', function(err, data, res) {});
+ * // POST http://httptest.cnodejs.net
+ * var args = { type: 'post', data: { foo: 'bar' } };
+ * urllib.request('http://httptest.cnodejs.net/test/post', args, function(err, data, res) {});
+ *
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ * @param options Optional, @see RequestOptions.
+ */
+export function request(url: string | url.URL, options?: RequestOptions): Promise<any>;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ * @param callback @see Callback
+ */
+export function request(url: string | url.URL, callback: Callback): void;
+/**
+ * @param url The URL to request, either a String or a Object that return by url.parse.
+ * @param options Optional, @see RequestOptions.
```

Comment:
This jsdoc is redundant -- the TypeScript already says `options: RequestOptions` and `callback: Callback`. Also, if `options` is optional, it should presumably be typed `options: RequestOptions | undefined`?